### DISCUSSION TITLE
Need Guidance to use the Publish_Release_To_GitHub task

### DISCUSSION THREAD

**Discussion by fslef** [gaelcolas/Sampler#467] [Q&A]
Hello,

I have configured my build.yaml to publish a GitHub release, but the task is being skipped during execution.
I suspect this issue may be due to missing prerequisites.

Here is the logs:

``` powershell
./build.ps1 -tasks publish                   
[pre-build] Starting Build Init
[build] Starting build with InvokeBuild.
[build] Parsing defined tasks
[build] Loading Configuration from /Users/<username>/gitrepos/fslef-cybersocforge/build.yaml
Importing tasks from module Sampler
         Loading Build-Module.ModuleBuilder.build.Sampler.ib.tasks...
         Loading ChocolateyPackage.build.Sampler.ib.tasks...
         Loading Clean.ModuleBuilder.build.Sampler.ib.tasks...
         Loading Create_Changelog_Branch.build.Sampler.ib.tasks...
         Loading Create_Release_Git_Tag.build.Sampler.ib.tasks...
         Loading DeployAll.PSDeploy.build.Sampler.ib.tasks...
         Loading DscResource.Test.build.Sampler.ib.tasks...
         Loading generateHelp.PlatyPS.build.Sampler.ib.tasks...
         Loading GuestConfig.build.Sampler.ib.tasks...
         Loading Invoke-Pester.pester.build.Sampler.ib.tasks...
         Loading JaCoCo.coverage.build.Sampler.ib.tasks...
         Loading release.module.build.Sampler.ib.tasks...
         Loading Set-SamplerTaskVariable...
         Loading SetPsModulePath.build.Sampler.ib.tasks...
Importing tasks from module Sampler.GitHubTasks
         Loading New-Release.GitHub.build.Sampler.GitHubTasks.ib.tasks...
Adding Workflow from configuration:
  +-> .
  +-> build
  +-> pack
  +-> test
  +-> publish
[build] Executing requested workflow: publish
Build publish /Users/<username>/gitrepos/fslef-cybersocforge/build.ps1
Redefined task '.'.
Task /publish/Publish_Release_To_GitHub skipped.
Done /publish 00:00:00.0001550
Build succeeded. 1 tasks, 0 errors, 0 warnings 00:00:00.1162080
```

Despite searching the documentation for this specific task, I have not found any relevant information. Could you please guide me to the appropriate documentation for this task or provide some troubleshooting guidance?

Thank you very much in advance.
François

**Comment by fslef** [comment]
.

**Reply by phatmandrake** [reply]
@fslef 

I think I'm running into the same issue! Were you able to resolve this?

**Comment by danubie** [comment]
Maybe you are missing one of those:
- variable Githubtoken
- Module PowerShellForGitHub

**Comment by johlju** [comment]
Look at what tasks the metatask `publish` runs, probably `Publish_release_to_GitHub` which is part of the module Sampler.GitHubTask. First make sure that module is in RequiredModules. Then make sure the requirements as @danubie mentioned exist:

https://github.com/gaelcolas/Sampler.GitHubTasks/blob/1f288c3bc35ab124ff9f8a8f405111c10761744e/.build/tasks/New-Release.GitHub.build.ps1#L54 

GitHub token must be a variable in the current session or as an environment variable, then this will resolve `GitHubToken` requirement.

https://github.com/gaelcolas/Sampler.GitHubTasks/blob/1f288c3bc35ab124ff9f8a8f405111c10761744e/.build/tasks/New-Release.GitHub.build.ps1#L27

**Reply by danubie** [reply]
I'm currently working to create a Github Workflow to use Sampler (BTW Thanks for creating and maintaining it!).
During this, I stumbled across another possible pitfall in presenting the Githubtoken to the task:

- In the env context of the job I defined an environment variable: GithubToken: "${{ secrets.GITHUB_TOKEN }}"
- You end up with an environment variable 'GITHUBTOKEN' (all uppercase) in the task
But
- Parameters to the task are passed through $GitHubToken = (property GitHubToken '')
- Inside Invoke-Builds, Get-buildPropery uses [Environment]::GetEnvironmentVariable
  [Environment]::GetEnvironmentVariable is case-sensitive on linux .NET Core!

If you now are **using a linux runner**, the environment variable will not be read, because of the different casing. The task would expect '$ENV:GithubToken' but gets $ENV:GITHUBTOKEN'.

Allthough everything seems to be prepared for Publish_Release_To_GitHub, it will result in "step Publish_Release_To_GitHub is skipped".
If you agree in this: Where to file an issue? Sampler or Invoke-Build?

**Reply by johlju** [reply]
This seems to be an issue with `property` in Invoke-Build module so I suggest to add an issue there. I have not seen this yet, because in Azure Pipelines you get the same casing for name of the variable. If adding the name GitHubToken as a secret variable name, that will be that exact casing on the hosted (build) agents. 

So if `property` would `.Upper()` all environment variables it would break Azure Pipelines on Linux and macOS runners. My guess is Invoke-Build might need to check if it runs in what environment and handle environment variable names respectively. 🤔

**Reply by danubie** [reply]
Thanks for your guidance. For now I switched to a Windows runner. There it seems to behave like in Azure Pipelines.
